marshmallow/fields.py

```diff
@@ -1195,31 +1195,23 @@ def _deserialize(self, value, attr, data):
         if not self.value_container and not self.key_container:
             return value
 
-        errors = {}
+        errors = collections.defaultdict(dict)
         values = list(value.values())
         keys = list(value.keys())
         if self.key_container:
             for idx, key in enumerate(keys):
                 try:
                     keys[idx] = self.key_container.deserialize(key)
                 except ValidationError as e:
-                    errors[key] = [
-                        'Invalid key: {}'.format(message)
-                        for message in e.messages
-                    ]
+                    errors[key]['key'] = e.messages
```

Comment:
~~I don't think changing the error structure is necessary to fix this bug. From what I can tell, the dict deserialization error structure I implemented is closer to the list deserialization structure than what was proposed in https://github.com/marshmallow-code/marshmallow/issues/730#issuecomment-361362161.~~

I actually like how using a dictionary simplifies the implementation by avoiding array extension and text prefixes, ~~but I propose moving changes to the error structure to a separate issue/PR~~.

Reply:
I see why it is needed now. Thanks for digging into this @lafrech.

Reply:
It might be helpful to provide a more technical insight about this fix in the PR description for subsequent reviewers. It wasn't clear from the conversation in the issue why this was a necessary.